diff --git a/library/asn1write.c b/library/asn1write.c
index 053dbb669..dd050e4b8 100644
--- a/library/asn1write.c
+++ b/library/asn1write.c
@@ -34,6 +34,10 @@
 #define mbedtls_free       free
 #endif
 
+#ifdef _LINUX_KERNEL_
+#define SIZE_MAX	(18446744073709551615UL)
+#endif
+
 int mbedtls_asn1_write_len( unsigned char **p, const unsigned char *start, size_t len )
 {
     if( len < 0x80 )
diff --git a/library/bignum.c b/library/bignum.c
index 8717c8abc..a52bf7227 100644
--- a/library/bignum.c
+++ b/library/bignum.c
@@ -1435,7 +1435,8 @@ int mbedtls_mpi_mul_mpi( mbedtls_mpi *X, const mbedtls_mpi *A, const mbedtls_mpi
     MBEDTLS_MPI_CHK( mbedtls_mpi_grow( X, i + j ) );
     MBEDTLS_MPI_CHK( mbedtls_mpi_lset( X, 0 ) );
 
-    for( size_t k = 0; k < j; k++ )
+    size_t k;
+    for( k = 0; k < j; k++ )
     {
         /* We know that there cannot be any carry-out since we're
          * iterating from bottom to top. */
@@ -1876,7 +1877,8 @@ static void mpi_montmul( mbedtls_mpi *A, const mbedtls_mpi *B, const mbedtls_mpi
     n = N->n;
     m = ( B->n < n ) ? B->n : n;
 
-    for( size_t i = 0; i < n; i++ )
+    size_t i;
+    for( i = 0; i < n; i++ )
     {
         mbedtls_mpi_uint u0, u1;
 
@@ -1952,7 +1954,8 @@ static int mpi_select( mbedtls_mpi *R, const mbedtls_mpi *T, size_t T_size, size
 {
     int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
 
-    for( size_t i = 0; i < T_size; i++ )
+    size_t i;
+    for( i = 0; i < T_size; i++ )
     {
         MBEDTLS_MPI_CHK( mbedtls_mpi_safe_cond_assign( R, &T[i],
                         (unsigned char) mbedtls_ct_size_bool_eq( i, idx ) ) );
diff --git a/library/ctr_drbg.c b/library/ctr_drbg.c
index 43f490e83..d03be5284 100644
--- a/library/ctr_drbg.c
+++ b/library/ctr_drbg.c
@@ -45,6 +45,11 @@
 #endif /* MBEDTLS_PLATFORM_C */
 #endif /* MBEDTLS_SELF_TEST */
 
+#ifdef _LINUX_KERNEL_
+#define SIZE_MAX	(18446744073709551615UL)
+#define INT_MAX		(2147483647)
+#endif
+
 /*
  * CTR_DRBG context initialization
  */
diff --git a/library/ecjpake.c b/library/ecjpake.c
index d467a6540..d17544200 100644
--- a/library/ecjpake.c
+++ b/library/ecjpake.c
@@ -970,7 +970,8 @@ static int self_test_rng( void *ctx, unsigned char *out, size_t len )
 
     (void) ctx;
 
-    for( size_t i = 0; i < len; i++ )
+    size_t i;
+    for( i = 0; i < len; i++ )
     {
         state = state * 1664525u + 1013904223u;
         out[i] = (unsigned char) state;
diff --git a/library/ecp.c b/library/ecp.c
index 67c46f2c3..4a560a199 100644
--- a/library/ecp.c
+++ b/library/ecp.c
@@ -3391,7 +3391,8 @@ static int self_test_rng( void *ctx, unsigned char *out, size_t len )
 
     (void) ctx;
 
-    for( size_t i = 0; i < len; i++ )
+    size_t i;
+    for( i = 0; i < len; i++ )
     {
         state = state * 1664525u + 1013904223u;
         out[i] = (unsigned char) state;
diff --git a/library/nist_kw.c b/library/nist_kw.c
index 1aea0b634..7a93b6717 100644
--- a/library/nist_kw.c
+++ b/library/nist_kw.c
@@ -48,6 +48,10 @@
 #endif /* MBEDTLS_PLATFORM_C */
 #endif /* MBEDTLS_SELF_TEST && MBEDTLS_AES_C */
 
+#ifdef _LINUX_KERNEL_
+#define SIZE_MAX	(18446744073709551615UL)
+#endif
+
 #if !defined(MBEDTLS_NIST_KW_ALT)
 
 #define KW_SEMIBLOCK_LENGTH    8
diff --git a/library/pk.c b/library/pk.c
index 1eab65a27..29425c8b7 100644
--- a/library/pk.c
+++ b/library/pk.c
@@ -50,6 +50,10 @@
 #define PK_VALIDATE( cond )        \
     MBEDTLS_INTERNAL_VALIDATE( cond )
 
+#ifdef _LINUX_KERNEL_
+#define SIZE_MAX	(18446744073709551615UL)
+#endif
+
 /*
  * Initialise a mbedtls_pk_context
  */
diff --git a/library/pk_wrap.c b/library/pk_wrap.c
index e234a8de3..afb06c771 100644
--- a/library/pk_wrap.c
+++ b/library/pk_wrap.c
@@ -19,6 +19,10 @@
 
 #include "common.h"
 
+#ifdef _LINUX_KERNEL_
+#define SIZE_MAX	(18446744073709551615UL)
+#endif
+
 #if defined(MBEDTLS_PK_C)
 #include "pk_wrap.h"
 #include "mbedtls/error.h"
diff --git a/library/psa_crypto.c b/library/psa_crypto.c
index b0116ddfb..729614d13 100644
--- a/library/psa_crypto.c
+++ b/library/psa_crypto.c
@@ -20,6 +20,10 @@
 
 #include "common.h"
 
+#ifdef _LINUX_KERNEL_
+#define SIZE_MAX	(18446744073709551615UL)
+#endif
+
 #if defined(MBEDTLS_PSA_CRYPTO_C)
 
 #if defined(MBEDTLS_PSA_CRYPTO_CONFIG)
diff --git a/library/psa_crypto_rsa.c b/library/psa_crypto_rsa.c
index 68f4bf16d..4610ba7c1 100644
--- a/library/psa_crypto_rsa.c
+++ b/library/psa_crypto_rsa.c
@@ -20,6 +20,10 @@
 
 #include "common.h"
 
+#ifdef _LINUX_KERNEL_
+#define SIZE_MAX	(18446744073709551615UL)
+#endif
+
 #if defined(MBEDTLS_PSA_CRYPTO_C)
 
 #include <psa/crypto.h>
